nmpolicy/internal/capture/resolver.go

```diff
@@ -40,52 +40,58 @@ type Parser interface {
 
 type Resolver interface {
 	Resolve(astPool map[string]ast.Node, state []byte) (map[string]types.CaptureState, error)
+	ResolveCaptureEntryPath(captureEntryPath ast.Node) (interface{}, error)
 }
 
-func New(leXer Lexer, parser Parser, resolver Resolver) Capture {
-	return Capture{
+func NewResolver(leXer Lexer, parser Parser, resolver Resolver) CaptureResolver {
+	return CaptureResolver{
 		lexer:    leXer,
 		parser:   parser,
 		resolver: resolver,
 	}
 }
 
-func (c Capture) Resolve(
-	capturesExpr map[string]string,
-	capturesCache map[string]types.CaptureState,
-	state []byte) (map[string]types.CaptureState, error) {
-	if len(capturesExpr) == 0 || len(state) == 0 && len(capturesCache) == 0 {
+func (c CaptureResolver) Resolve(
+	captureExpressions map[string]string,
+	capturedStateCache map[string]types.CaptureState,
+	state []byte) (*ResolvedCapture, error) {
+	if len(captureExpressions) == 0 || len(state) == 0 && len(capturedStateCache) == 0 {
 		return nil, nil
 	}
 
-	capturesState := filterCacheBasedOnExprCaptures(capturesCache, capturesExpr)
-	capturesExpr = filterOutExprBasedOnCachedCaptures(capturesExpr, capturesCache)
+	filteredCapturedStateCache := filterCacheBasedOnExprCaptures(capturedStateCache, captureExpressions)
+	filteredCaptureExpressions := filterOutExprBasedOnCachedCaptures(captureExpressions, filteredCapturedStateCache)
 
-	astPool := map[string]ast.Node{}
-	for capID, capExpr := range capturesExpr {
-		tokens, err := c.lexer.Lex(capExpr)
+	captureASTPool := map[string]ast.Node{}
+	for captureEntryName, captureEntryExpression := range filteredCaptureExpressions {
+		captureEntryTokens, err := c.lexer.Lex(captureEntryExpression)
 		if err != nil {
 			return nil, fmt.Errorf("failed to resolve capture expression, err: %v", err)
 		}
 
-		astRoot, err := c.parser.Parse(tokens)
+		captureEntryAST, err := c.parser.Parse(captureEntryTokens)
 		if err != nil {
 			return nil, fmt.Errorf("failed to resolve capture expression, err: %v", err)
 		}
 
-		astPool[capID] = astRoot
+		captureASTPool[captureEntryName] = captureEntryAST
 	}
 
-	resolvedCapsState, err := c.resolver.Resolve(astPool, state)
+	capturedState, err := c.resolver.Resolve(captureASTPool, state)
 	if err != nil {
 		return nil, fmt.Errorf("failed to resolve capture expression, err: %v", err)
 	}
 
-	for capID, capExpr := range resolvedCapsState {
-		capturesState[capID] = capExpr
+	for captureEntryName, filteredCapturedStateCacheEntry := range filteredCapturedStateCache {
+		capturedState[captureEntryName] = filteredCapturedStateCacheEntry
 	}
 
-	return capturesState, nil
+	return &ResolvedCapture{
+		lexer:         c.lexer,
```

Comment:
Same here, I don't understand why the resolve should return anything besides the states. The `lexer`, `parser` and `resolver` are stateless, there is no reason to reuse them.

Reply:
We need them again, and `resolver` is not statedless.